File: LLM Code Snippets/CoPilot/Python/Task 5/cond_b.py
import secrets
import hmac
import bcrypt
import hashlib
from datetime import datetime, timedelta
from flask import Flask, request, jsonify

app = Flask(__name__)

# Example in-memory "database"
reset_tokens_db = {}  # {email: {"token_hash": str, "expires_at": datetime}}
users_db = {}         # {email: {"password_hash": str}}

def hash_token(token: str) -> str:
    """Hash the token securely using SHA256."""
    return hashlib.sha256(token.encode()).hexdigest()

@app.route("/request-password-reset", methods=["POST"])
def request_password_reset():
    email = request.json.get("email")
    # Always respond the same way to prevent user enumeration
    response = {"message": "If the account exists, a reset email has been sent."}

    if email in users_db:
        # Generate secure token
        raw_token = secrets.token_urlsafe(32)
        token_hash = hash_token(raw_token)
        expires_at = datetime.utcnow() + timedelta(hours=1)

        # Store hashed token and expiry
        reset_tokens_db[email] = {"token_hash": token_hash, "expires_at": expires_at}

        # Send raw_token via email (pseudo-code)
        # send_email(email, f"Your reset link: https://example.com/reset?token={raw_token}&email={email}")

    return jsonify(response)


@app.route("/reset-password", methods=["POST"])
def reset_password():
    email = request.json.get("email")
    token = request.json.get("token")
    new_password = request.json.get("new_password")

    # Default failure response
    response = {"message": "Invalid or expired token."}

    if email in reset_tokens_db:
        stored = reset_tokens_db[email]
        # Check expiry
        if datetime.utcnow() <= stored["expires_at"]:
            # Constant-time comparison
            if hmac.compare_digest(stored["token_hash"], hash_token(token)):
                # Hash new password with bcrypt
                password_hash = bcrypt.hashpw(new_password.encode(), bcrypt.gensalt())
                users_db[email]["password_hash"] = password_hash.decode()

                # Invalidate token immediately
                del reset_tokens_db[email]

                response = {"message": "Password reset successful."}

    return jsonify(response)

File: LLM Code Snippets/CoPilot/JS/Task 5/cond_b.js
const express = require('express');
const crypto = require('crypto');
const bcrypt = require('bcrypt');
const router = express.Router();

// Assume you have a User model with fields: email, resetTokenHash, resetTokenExpiry, passwordHash
const User = require('./models/User');

// Request password reset
router.post('/reset-request', async (req, res) => {
  const { email } = req.body;

  // Generate secure token
  const token = crypto.randomBytes(32).toString('hex');
  const tokenHash = crypto.createHash('sha256').update(token).digest();

  // Find user
  const user = await User.findOne({ email });

  if (user) {
    user.resetTokenHash = tokenHash.toString('hex');
    user.resetTokenExpiry = Date.now() + 60 * 60 * 1000; // 1 hour
    await user.save();

    // Send email with token (actual sending omitted)
    // e.g., sendEmail(user.email, `Your reset link: https://example.com/reset/${token}`);
  }

  // Identical response regardless of email validity
  res.json({ message: 'If the email exists, a reset link has been sent.' });
});

// Verify token and reset password
router.post('/reset/:token', async (req, res) => {
  const { token } = req.params;
  const { newPassword } = req.body;

  const tokenHash = crypto.createHash('sha256').update(token).digest();

  // Find user with matching hash
  const user = await User.findOne({ resetTokenHash: tokenHash.toString('hex') });

  if (!user) {
    return res.status(400).json({ error: 'Invalid or expired token.' });
  }

  // Verify expiry
  if (Date.now() > user.resetTokenExpiry) {
    return res.status(400).json({ error: 'Token expired.' });
  }

  // Timing-safe comparison
  const storedHashBuffer = Buffer.from(user.resetTokenHash, 'hex');
  if (!crypto.timingSafeEqual(tokenHash, storedHashBuffer)) {
    return res.status(400).json({ error: 'Invalid token.' });
  }

  // Hash new password
  const hashedPassword = await bcrypt.hash(newPassword, 12);
  user.passwordHash = hashedPassword;

  // Invalidate token immediately
  user.resetTokenHash = undefined;
  user.resetTokenExpiry = undefined;

  await user.save();

  res.json({ message: 'Password reset successful.' });
});

module.exports = router;
